ros2action/ros2action/api/__init__.py

```diff
@@ -139,3 +142,29 @@ def __call__(self, prefix, parsed_args, **kwargs):
                 if n == action_name:
                     return t
         return []
+
+
+class ActionGoalPrototypeCompleter:
+    """Callable returning an action goal prototype."""
+
+    def __init__(self, *, action_type_key=None):
+        self.action_type_key = action_type_key
+
+    def __call__(self, prefix, parsed_args, **kwargs):
+        full_action_type = getattr(parsed_args, self.action_type_key)
+        try:
+            parts = full_action_type.split('/')
+            if len(parts) == 1:
+                raise ValueError()
+            if len(parts) == 2:
+                parts = [parts[0], 'action', parts[1]]
+            if not all(parts):
+                raise ValueError()
+            package_name = parts[0]
+            interface_type = parts[1]
+            action_type = parts[-1]
+        except ValueError:
+            raise RuntimeError('The passed action type is invalid')
+        action = import_message_from_namespaced_type(
+            NamespacedType([package_name, interface_type], action_type))
```

Comment:
Why is the type first split, three parts explicitly extracted and then put back together? I would be good to rewrite this part to work for any number of namespaces (e.g. more than 2).

Also as noted in various other cases where this logic is present in this repo it should be considered to not copy-n-paste another instance / variation of it her but to put a generic function in `rosidl_runtime_py` instead.

Reply:
In the case of an action, the second item of the namespace should always be `action`, right?
e.g.: `std_msgs/msg/String` shouldn't be accepted as a valid action.
We are not checking that in other places of this repo (see https://github.com/ros2/ros2cli/pull/304#discussion_r308410962).

Reply:
Do you mean `package_name, interface_type, action_type` ? Those variables exists here mostly for readability. Notice that they are not fully put back together. I can remove them if you prefer.

How would it work with more than 2 namespaces ? What would be those other namespaces and what would the pattern be ? I could try to put a generic function together with those info.

Reply:
@ivanpauno I was actually thinking of adding a check on `action`, was surprised there was none in [send_goal](https://github.com/ros2/ros2cli/blob/master/ros2action/ros2action/verb/send_goal.py#L84).
I'll do so for this chunk of code in this PR and consider another PR for `send_goal`.

Reply:
`NamespacedType` takes two arguments: the first one are the namespaces in a list - the list can have any length, the second is the name of the type. To from the `split()` point of view `[-1]` becomes the second argument and `[0:-1]` the first argument.

Injecting `action` when the split length is only 2 is syntactic sugar for backward compatibility.

Reply:
At the moment the second part is `action`. But there is no reason to limit the type name to that convention. Therefore I don't think it should be enforced here.

Reply:
> At the moment the second part is `action`. But there is no reason to limit the type name to that convention. Therefore I don't think it should be enforced here.

@dirk-thomas I don't understand why you ':+1:' @ivanpauno comment then. From the comment linked it seems like to good idea to enforce `action` here.

Reply:
@dirk-thomas I removed `package_name, interface_type, action_type` and called directly `NamespacedType(parts[0:-1], parts[-1])`.

Reply:
There is a big difference between "the second item of the namespace should be `action`" and "enforcing that the second item is `action`". As I said above: atm the second item is `action` for all defined interfaces but that might not we the case in the future when we allow more flexibility / users to diverge from conventions.

Reply:
There sure is, but then it makes it difficult not to accept e.g. `std_msgs/msg/String` or whatnot.

Reply:
> There sure is, but then it makes it difficult not to accept e.g. `std_msgs/msg/String` or whatnot.

You could check the returned instance and check if it looks like an action or not?

Reply:
Something more explicit than `return [message_to_yaml(action.Goal())]` ?  
`if not hasattr(action, 'Goal'): raise RuntimeError('Expected an action')` ?

Reply:
Updated with a `hasattr(action, 'Goal')` check.